File: app/src/main/java/com/kksionek/gdzietentramwaj/di/AppModule.java
package com.kksionek.gdzietentramwaj.di;

import android.arch.persistence.room.Room;
import android.content.Context;

import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import com.kksionek.gdzietentramwaj.DataSource.Room.MyDatabase;
import com.kksionek.gdzietentramwaj.DataSource.Room.TramDao;
import com.kksionek.gdzietentramwaj.DataSource.TramInterface;
import com.kksionek.gdzietentramwaj.view.AdProvider;
import com.kksionek.gdzietentramwaj.view.AdProviderInterface;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.reactivex.schedulers.Schedulers;
import okhttp3.OkHttpClient;
import retrofit2.CallAdapter;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

@Module
public class AppModule {

    private final Context mContext;

    public AppModule(Context context) {
        mContext = context;
    }

    @Singleton @Provides
    Context provideContext() {
        return mContext;
    }

    @Singleton @Provides
    MyDatabase getMyDatabase(Context context) {
        return Room.databaseBuilder(context, MyDatabase.class, "favorites.db")
                .addMigrations(MyDatabase.MIGRATION_1_2, MyDatabase.MIGRATION_2_3)
                .build();
    }

    @Singleton @Provides
    TramDao getTramDao(MyDatabase myDatabase) {
        return myDatabase.tramDao();
    }

    @Singleton @Provides
    TramInterface provideTramInterface(OkHttpClient okHttpClient, CallAdapter.Factory rxAdapter) {
        return new Retrofit.Builder()
                .addConverterFactory(GsonConverterFactory.create())
                .client(okHttpClient)
                .addCallAdapterFactory(rxAdapter)
                .baseUrl("https://api.um.warszawa.pl/")
                .build().create(TramInterface.class);
    }

    @Singleton @Provides
    OkHttpClient provideOkHttpClient() {
        // DEBUG
//        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
//        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        return new OkHttpClient.Builder()
//                .addInterceptor(interceptor)
                .build();
    }

    @Singleton @Provides
    CallAdapter.Factory getCallAdapterFactory() {
        return RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io());
    }

    @Singleton @Provides
    AdProviderInterface provideAdProvider() {
        return new AdProvider();
    }
}

File: app/src/main/java/com/kksionek/gdzietentramwaj/di/AppModule.kt
package com.kksionek.gdzietentramwaj.di

import android.arch.persistence.room.Room
import android.content.Context
import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import com.kksionek.gdzietentramwaj.CrashReportingService
import com.kksionek.gdzietentramwaj.CrashlyticsCrashReportingService
import com.kksionek.gdzietentramwaj.TramApplication
import com.kksionek.gdzietentramwaj.dataSource.room.MyDatabase
import com.kksionek.gdzietentramwaj.dataSource.room.TramDao
import dagger.Module
import dagger.Provides
import io.reactivex.schedulers.Schedulers
import retrofit2.CallAdapter
import javax.inject.Singleton

@Module
class AppModule(private val application: TramApplication) {

    @Singleton
    @Provides
    internal fun provideCallAdapterFactory(): CallAdapter.Factory =
        RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io())

    @Singleton
    @Provides
    internal fun provideApplication(): TramApplication = application

    @Singleton
    @Provides
    internal fun provideContext(): Context = application

    @Singleton
    @Provides
    internal fun provideCrashReportingService(): CrashReportingService =
        CrashlyticsCrashReportingService()

    @Singleton
    @Provides
    internal fun getMyDatabase(context: Context): MyDatabase {
        return Room.databaseBuilder(context, MyDatabase::class.java, "favorites.db")
            .addMigrations(
                MyDatabase.MIGRATION_1_2,
                MyDatabase.MIGRATION_2_3,
                MyDatabase.MIGRATION_3_4
            )
            .build()
    }

    @Singleton
    @Provides
    internal fun getTramDao(myDatabase: MyDatabase): TramDao = myDatabase.tramDao()
}
